waspc/src/Wasp/Project/Db.hs

```diff
@@ -1,23 +1,48 @@
 module Wasp.Project.Db
   ( makeDevDatabaseUrl,
     databaseUrlEnvVarName,
+    getDbSystemFromPrismaSchema,
   )
 where
 
 import StrongPath (Abs, Dir, Path')
 import qualified Wasp.AppSpec as AS
-import qualified Wasp.AppSpec.App as AS.App
 import qualified Wasp.AppSpec.App.Db as AS.App.Db
+import qualified Wasp.AppSpec.App.Db as AS.Db
 import Wasp.Project.Common (WaspProjectDir)
 import qualified Wasp.Project.Db.Dev.Postgres as DevPostgres
+import qualified Wasp.Psl.Ast.Schema as Psl.Ast
+import Wasp.Psl.Util (findPrismaConfigBlockKeyValue)
 
-makeDevDatabaseUrl :: Path' Abs (Dir WaspProjectDir) -> [AS.Decl] -> Maybe String
-makeDevDatabaseUrl waspProjectDir decls = do
-  (appName, app) <- AS.getApp decls
-  dbSystem <- AS.App.Db.system =<< AS.App.db app
+makeDevDatabaseUrl ::
+  Path' Abs (Dir WaspProjectDir) ->
+  AS.Db.DbSystem ->
+  [AS.Decl] ->
+  Maybe String
+makeDevDatabaseUrl waspProjectDir dbSystem decls = do
+  (appName, _) <- AS.getApp decls
   case dbSystem of
-    AS.App.Db.SQLite -> Nothing
     AS.App.Db.PostgreSQL -> Just $ DevPostgres.makeDevConnectionUrl waspProjectDir appName
+    _allOtherCases -> Nothing
 
 databaseUrlEnvVarName :: String
 databaseUrlEnvVarName = "DATABASE_URL"
+
+getDbSystemFromPrismaSchema :: Psl.Ast.Schema -> AS.Db.DbSystem
+getDbSystemFromPrismaSchema prismaSchema =
+  case getProviderFromPrismaSchema prismaSchema of
+    -- We parse raw config block values from Prisma file,
+    -- so we need match the provider names with quotes.
+    Just "\"postgresql\"" -> AS.App.Db.PostgreSQL
+    Just "\"sqlite\"" -> AS.App.Db.SQLite
+    Just provider -> AS.App.Db.UnsupportedDbSystem provider
+    Nothing -> AS.App.Db.MissingDbSystem
+
+getProviderFromPrismaSchema :: Psl.Ast.Schema -> Maybe String
+getProviderFromPrismaSchema =
+  findPrismaConfigBlockKeyValue "provider"
+    -- As per Prisma's docs there can be only ONE datasource block in the schema.
```

Comment:
I'm not sure how to enforce this with Parsec, do you maybe have some tips on how to ensure there is only _at max 1 datasource_ in the parsed AST?

Reply:
One way is to track state in Parsec -> you could, in datasource parser, check first if you already parsed one, in the state. If yes, you throw error. If not, you parse it and add to state that you parsed it.
In Parsec, `ParsecT s u m a`, this `u` is user state, so your state you want to drag around with parser (user state). So you could make a `data PslParserState = PslParserState { didParseDatasource :: Bool }`, pass that as `s` param to Parsec type, and then use `getState`, `putState`, `modifyState` to manipulate that.

Another way is to allow parsing of multiple datasources, but then have another type of AST, e.g. ValidAst, that can hve only one datasource by design, and you would have function `validate :: Ast -> Either [Error] ValidAst`. This can be tricky though, in the sense that you have to basically write another Ast and that is a lot of duplication. There are some sophisticated methods to actually use just one AST, but that is quite complex and not sometihng we want to get into.

Finally, you can do what we did with AppSpec -> have Valid.hs file which has function that validates AppSpec, and then also get getters which say in comments that they expect they work on validated AppSpec. Bad thing here is this is not enforced by compiler, so you have to be careful to use these only on validated AppSpec, but still, in practice it doesn't work so badly and you don't have to do that crazy complex machinery I mentioned right above.

Here, I would go with first or last suggestion.

Reply:
I suppose I already implemented the last suggestion? Where we parsed all the datasources and then validate the number of them in the `Valid.hs`?

Reply:
Right, but is that in the Valid.hs for Prisma schema, or for AppSpec? The idea is you have it at the Prisma Schema (PSL) level, similar like we have for AppSpec.

Reply:
Okay, now we have Prisma schema validation as a separate concept from the AppSpec validation, making it look a lot cleaner and explicit IMHO. Nice suggestion 👍

Reply:
Awesome!